### DISCUSSION TITLE
Welcome to CS Tools Discussions!

### DISCUSSION THREAD

**Discussion by boonhapus** [thoughtspot/cs_tools#55] [General]
## 👋 Welcome!

We’re using GitHub Discussions as a place to connect with other members of our community. 

**We hope that you..**
  🤗 **Welcome others** and are open-minded.
  🤔 **Ask questions** you’re wondering about.
  🤚 **Get help** with Install & Configuration.
  💡 **Share ideas** of how you're using or want to use the tools.
  🧑‍🤝‍🧑 **Engage other** members in the community.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.
  
  When you post a **[New Discussion](https://github.com/thoughtspot/cs_tools/discussions/new/choose)** in our community, all the maintainers of this library will get notified!
    
![image](https://github.com/thoughtspot/cs_tools/assets/14192437/3bf7984e-244d-4b71-9d6a-e2f60811d8dc)

**Comment by girishbhagwani** [comment]
Hi Team,

I am unable to understand what should be the format of CSV that I should pass to CSTools sync users.
I am using the below command for syncing my users but at the moment, the CSV file is blank. I am unable to find any relevant example as well. Please guide me here. 

Command :-
`cs_tools tools user-management sync --users "C:\Users\...\AppData\Roaming\cs_tools\user" --syncer csv://C:\Users\...\AppData\Roaming\cs_tools\definition.toml --dry-run`

**definition.toml**

```
[configuration]
directory = 'C:\Users\...\AppData\Roaming\cs_tools'
delimiter = ','
escape_character = '\'
zipped = false
```

**Reply by boonhapus** [reply]
@girishbhagwani try doing this one first.

`cs_tools tools user-management sync --syncer csv://directory=C:/Users/... --export`

It will export the data to the `C:/User/...` directory so you can see the files and their structure. If you want to update just the users, then you would change just the file called `C:/Users/.../ts_auth_sync_users.csv`.

The `--users` , `--groups` , and `--associations` options in this command are called "directives", and they simply tell the Syncer where to find certain datapoints. For CSV, which stores table-like files in a directory, the directive would tell it where to find the users CSV. If we used the Snowflake syncer, it would denote a Table instead.

By default, we'll `--export` with the same directive name.

**Reply by girishbhagwani** [reply]
Hi Team,

I am facing this issue while pussing this command 

`cs_tools tools user-management sync --syncer csv://directory=C:\Users\c35045\AppData\Roaming\cs_tools --new-user-password pass@321 --verbose --apply-changes`


Error
`[\"Error Code: FAILED_TO_COMMIT Incident Id: d5c2a1a2-dee5-4a6e-8f6e-734243afe608\\nError Message: com.thoughtspot.atlas.AtlasException:  code\\u003dVERSION_CONFLICT, Node Content updated before put\\u003d7ce5b099-bd33-4642-b1e4-d3944a42be5f\"]"}`


`Log file`
...httpx.HTTPStatusError: Server error '500 ' for url 'https://askme.rblbank.com/callosum/v1/tspublic/v1/user/sync'
For more information check: https://httpstatuses.com/500

**Reply by boonhapus** [reply]
👋 @girishbhagwani - we discussed this one live in a call, but I will answer here as well for others who find it.

The __User Management__ sync command wraps the `user/sync` ThoughtSpot API. When you use this tool, you're essentially telling ThoughtSpot "your user and groups should look like this", and then some work goes on behind the scenes to make that happen after the API returns. If you run this tool in quick succession, you could tell the internal database "hey, I want to update User XYZ" as the background process updates UserXYZ. This causes the `FAILED_TO_COMMIT: VERSION_CONFLICT` error. The error naturally resolves itself once the background process completes, and then you're able to run the next `sync` command.

In summary, if you see this error, you should take more time in between calling the tool to let the system catch up.

**Reply by girishbhagwani** [reply]
Thankyou soo much for giving your time and thankyou to the whole team who all are making there efforts to build such a beautiful tool!!

**Comment by bmofkn** [comment]
Has anyone been successful in using cs_tools on macOS using HomeBrew's Python?

cs_tools installs successfully but any attempt to run a cs_tools command results in the following error:

`File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/inspect.py", line 2722, in __init__
    raise ValueError('{!r} is not a valid parameter name'.format(name))
ValueError: 'not' is not a valid parameter name`

**Reply by boonhapus** [reply]
Well this is a new one @bmofkn! It appears this is a [__bug in__ `CPython -> pydantic -> fastapi`](https://github.com/tiangolo/fastapi/issues/5048#issuecomment-1170204100). I've reproduced this locally and have just pushed a fix. Can you run the bootstrapper reinstall command? This will grab the latest 1.4.7.

```shell
curl \
    --silent --show-error --location-trusted \
    https://raw.githubusercontent.com/thoughtspot/cs_tools/master/cs_tools/updater/_bootstrapper.py \
    | python3 - --reinstall 
```

**Reply by bmofkn** [reply]
@boonhapus , your "fix" worked!  Thanks for the quick reply!  
<img width="841" alt="cs_tools works" src="https://github.com/thoughtspot/cs_tools/assets/119008346/9d588668-5426-43e2-93b0-68dbfdb885e7">

**Comment by girishbhagwani** [comment]
Hi @boonhapus,

I am not able to use transfer ownership of users
command I am using
"`cs_tools tools user-management transfer --from 3...@r.....bank.in --to tsadmin`"

error
"`Client error '405 ' for url 'https://askme.rblbank.com/dataflow/diapi/transferofownership`'"

**Reply by boonhapus** [reply]
Hey @girishbhagwani , I think I've found the problem here. I'll add this into the next update within a few days!

**Reply by boonhapus** [reply]
Hi @girishbhagwani , this one should be fixed with the latest update. Please try `cs_tools self update` and let know if you run into any issues!

**Comment by suresh900** [comment]
Hi Team,
Can any one help us on this command, what happened when we run this command -
cs_tools tools user-management sync --syncer csv://directory=C:\Users\123.csv --config PROD1234 --apply-changes

**Comment by bmofkn** [comment]
Been loving cs_tools!  Use it every day to backup the TML for all of our objects and store in git.

We've recently encountered an issue with a few worksheets.  We use the following command to initiate the TML backup:
`cs_tools tools scriptability export --include-types worksheet ./ --config prod_ts`

Since Nov 1, 2023, we've been receiving the following error:
`thoughtspot_tml.exceptions.TMLDecodeError: WorksheetEDocProto.__init__() got an unexpected keyword argument 'action_object_associations'
supplied data does not produce a valid TML (Worksheet) document`

The worksheets that are failing all have this as the last few lines in their TML:
```
  action_object_associations:
  - action_name: Open Shipment
    context: CONTEXT_MENU
    enabled: true
```


We have implemented context menus for our TS Embedded users to quickly right-click a row or cell and have it open a window into our product.  

Question:  Did something change around the beginning of Nov that would suddenly cause those trailing lines to throw an error when using `cs_tools tools scriptability export --include-types worksheet ./ --config prod_ts` ?

**Reply by boonhapus** [reply]
Hey @bmofkn ! This is because now your TML has added those new lines. Only **[recently did we being to support this additional context](https://github.com/thoughtspot/thoughtspot_tml/commit/b8e86efffd84d5de21b2fbbb67bda9773b0e44af)** in the upstream `thoughtspot_tml` library. To fix this, you can activate the virtual environment and simply do a `pip install thoughtspot_tml --upgrade`. Use `cs_tools self info` to grab the path for the "activate virtualenv".

```terminal
(.venv) PS C:\Users\CooperNi\Downloads> cs_tools self info --anonymous
╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                                    │
│           Info snapshot taken on 2023-11-15                                                                        │
│                                                                                                                    │
│               CS Tools: 1.4.10                                                                                     │
│         Python Version: Python 3.11.6 (tags/v3.11.6:8b6ee5b, Oct  2 2023, 14:57:12) [MSC v.1935 64 bit (AMD64)]    │
│            System Info: Windows (detail: Windows-10-10.0.22621-SP0)                                                │
│      Configs Directory: C:\Users\ {anonymous} \AppData\Roaming\cs_tools                                            │
│    Activate VirtualEnv: C:\Users\ {anonymous} \Downloads\.venv\Scripts\Activate.ps1                                │
│          Platform Tags: win-amd64                                                                                  │
│                                                                                                                    │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

PS C:\Users\CooperNi> C:\Users\CooperNi\Downloads\.venv\Scripts\Activate.ps1

(.venv) PS C:\Users\CooperNi> pip install thoughtspot_tml --upgrade
```
